### PR TITLE
Prevent special chars in case search form

### LINKED ISSUE #3513: Breaking '\' in case filtering (assignees email)
```
2022-08-02 13:26:29,337 - [scout.server.app](http://scout.server.app/) - ERROR: Exception on /cust004/cases [GET] [in /venv/lib/python3.8/site-packages/flask/app.py:1455]
Traceback (most recent call last):
  File "/venv/lib/python3.8/site-packages/flask/app.py", line 2077, in wsgi_app
    response = self.full_dispatch_request()
  File "/venv/lib/python3.8/site-packages/flask/app.py", line 1525, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/venv/lib/python3.8/site-packages/flask_cors/extension.py", line 165, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/venv/lib/python3.8/site-packages/flask/app.py", line 1523, in full_dispatch_request
    rv = self.dispatch_request()
  File "/venv/lib/python3.8/site-packages/flask/app.py", line 1509, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/home/worker/app/scout/server/utils.py", line 85, in decorated_function
    context = func(*args, **kwargs)
  File "/home/worker/app/scout/server/blueprints/institutes/views.py", line 67, in cases
    return controllers.cases(store, request, institute_id)
  File "/home/worker/app/scout/server/blueprints/institutes/controllers.py", line 371, in cases
    all_cases = store.cases(
  File "/home/worker/app/scout/adapter/mongo/case.py", line 352, in cases
    order = self._populate_name_query(query, name_query, owner, collaborator)
  File "/home/worker/app/scout/adapter/mongo/case.py", line 239, in _populate_name_query
    query["assignees"] = {"$in": [user["email"] for user in users]}
  File "/home/worker/app/scout/adapter/mongo/case.py", line 239, in <listcomp>
    query["assignees"] = {"$in": [user["email"] for user in users]}
  File "/venv/lib/python3.8/site-packages/pymongo/cursor.py", line 1238, in next
    if len(self.__data) or self._refresh():
  File "/venv/lib/python3.8/site-packages/pymongo/cursor.py", line 1155, in _refresh
    self.__send_message(q)
  File "/venv/lib/python3.8/site-packages/pymongo/cursor.py", line 1044, in __send_message
    response = client._run_operation(
  File "/venv/lib/python3.8/site-packages/pymongo/mongo_client.py", line 1424, in _run_operation
    return self._retryable_read(
  File "/venv/lib/python3.8/site-packages/pymongo/mongo_client.py", line 1525, in _retryable_read
    return func(session, server, sock_info, secondary_ok)
  File "/venv/lib/python3.8/site-packages/pymongo/mongo_client.py", line 1420, in _cmd
    return server.run_operation(
  File "/venv/lib/python3.8/site-packages/pymongo/server.py", line 130, in run_operation
    _check_command_response(first, sock_info.max_wire_version)
  File "/venv/lib/python3.8/site-packages/pymongo/helpers.py", line 167, in _check_command_response
    raise OperationFailure(errmsg, code, response, max_wire_version)
pymongo.errors.OperationFailure: Regular expression is invalid: \ at end of pattern, full error: {'operationTime': Timestamp(1659439582, 1), 'ok': 0.0, 'errmsg': 'Regular expression is invalid: \\ at end of pattern', 'code': 51091, 'codeName': 'Location51091', '$clusterTime': {'clusterTime': Timestamp(1659439582, 1), 'signature': {'hash': b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', 'keyId': 0}}}
```

### PR DESCRIPTION
**Pull request by northwestwitch**
Fix #3513 

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the paxa software installed on `hasta`:
 - Log in into hasta: `ssh <USER.NAME>@hasta.scilifelab.se`
 - Activate the staging environment with the command `us` (Use Stage)
 - Run the command `paxa` and follow the instructions. Make sure to specify scout-stage as the resource you request and the server cg-vm1 as server.
1. Log out from the hasta server.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, log out from `cg-vm1` and log in again in the `hasta` server, repeat the `hasta` and `paxa` procedure, which will release the allocated resource (scout-stage) to be used for testing by other users.
</details>

**How to test**:
1. Locally or on cg-vm1
2. Go to an institute and try to type any of the following chars in the case search form: ^<>?!=\\/
3. The form should not validate and display an error
4. Try to reproduce the error from #3513 --> assignee's email search with `\    -` (doesn't matter how many spaces between the 2 characters)
5. The form should display error on submit, while on main branch the page should crash

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [ ] tests executed by

### CODE REVIEW

File: scout/server/blueprints/institutes/templates/overview/cases.html [hunk 1]
```diff
@@ -28,8 +28,8 @@
         {{ form.search_type(class="form-control") }}
       </div>
       <div class="col-4">
-        {{ form.search_term.label(class="control-label") }}
-        {{ form.search_term(class="form-control", placeholder='example:18201') }}
+          <label class="control-label" for="search_term">Search cases</label>
+          <input class="form-control" id="search_term" name="search_term" placeholder="example:18201" type="text" pattern="[^<>?!=\\/]*" title="Characters ^<>?!=\/ are not allowed">
```
Comment (dnil) [inline scout/server/blueprints/institutes/templates/overview/cases.html:32]:
```suggestion
          <input class="form-control" id="search_term" name="search_term" placeholder="example:18201" type="text" pattern="[^<>?!=\\/]*" data-bs-toggle="tooltip" title="Characters ^<>?!=\/ are not allowed">
```

**Review by dnil — APPROVED** [review]
I can reproduce the issue. This is a nice way to solve it, as it directly informs the user, but it doesn't quite go to the root of it.
It would also feel satisfying to fix the issue at
https://github.com/Clinical-Genomics/scout/blob/9b17661bfdf497c6606cdb2c18a89e142594eb90/scout/adapter/mongo/case.py#L234.

At any rate, as discussed, lets not forget to do the same on the dashboard! 😸

File: scout/server/blueprints/institutes/templates/overview/cases.html (hunk 1 above)
Comment (northwestwitch) [reply scout/server/blueprints/institutes/templates/overview/cases.html:32]: I tried locally and your suggestion doesn't display a tooltip. Why a tooltip BTW? This is form validation works fine already:

image

**Comment by northwestwitch** [comment]
> I can reproduce the issue. This is a nice way to solve it, as it directly informs the user, but it doesn't quite go to the root of it. It would also feel satisfying to fix the issue at
> 
> https://github.com/Clinical-Genomics/scout/blob/9b17661bfdf497c6606cdb2c18a89e142594eb90/scout/adapter/mongo/case.py#L234
> 
> .
> At any rate, as discussed, lets not forget to do the same on the dashboard! 😸

Fixed by introducing a re.escape to any search term passed to the query. Fixed also the form field in the dashboard page

File: scout/adapter/mongo/case.py [hunk 2]
```diff
@@ -233,7 +234,9 @@ def _populate_name_query(self, query, name_query, owner=None, collaborator=None)
         if query_field == "user":
             query_terms = query_term.split(" ")
             user_query = {
-                "$and": [{"name": {"$regex": term, "$options": "i"}} for term in query_terms]
+                "$and": [
+                    {"name": {"$regex": re.escape(term), "$options": "i"}} for term in query_terms
```
Comment (dnil) [inline scout/adapter/mongo/case.py:238]: 👍

File: scout/adapter/mongo/case.py [hunk 3]
```diff
@@ -156,7 +157,7 @@ def _populate_name_query(self, query, name_query, owner=None, collaborator=None)
         """
         order = None
         query_field = name_query.split(":")[0]  # example:status
-        query_term = name_query[name_query.index(":") + 1 :].strip()
+        query_term = re.escape(name_query[name_query.index(":") + 1 :].strip())
```
Comment (dnil) [inline scout/adapter/mongo/case.py:160]: 👍

File: scout/server/blueprints/institutes/templates/overview/cases.html (hunk 1 above)
Comment (northwestwitch) [reply scout/server/blueprints/institutes/templates/overview/cases.html:32]: Ahh OK, title is actually used by the HTML5 validation to display the error!

Comment (dnil) [reply scout/server/blueprints/institutes/templates/overview/cases.html:32]: It seems to be rather variable across user agents; see https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/pattern. Oh, one thing, the caret should not be in the title list - it is the NOT for the pattern match operation. 😊

**Comment by dnil** [comment]
For reference, a loooong hover with Safari will by default show the `title` as if a tooltip, but it does not make it part of the error message.

**Pull request merged**